nomad/host_volume_endpoint_test.go

```diff
@@ -0,0 +1,571 @@
+	vol4.Name = "foobaz-example"
+
+	registerReq.Volumes = []*structs.HostVolume{vol3, vol4}
+	registerReq.Namespace = ns2
+	err = msgpackrpc.CallWithCodec(codec, "HostVolume.Register", registerReq, &registerResp)
+	must.NoError(t, err)
+
+	testCases := []struct {
+		name         string
+		req          *structs.HostVolumeListRequest
+		expectVolIDs []string
+	}{
+		{
+			name: "wrong namespace for token",
+			req: &structs.HostVolumeListRequest{
+				QueryOptions: structs.QueryOptions{
+					Region:    srv.Region(),
+					Namespace: ns1,
+					AuthToken: otherToken,
+				},
+			},
+			expectVolIDs: []string{},
+		},
+		{
+			name: "query by namespace",
+			req: &structs.HostVolumeListRequest{
+				QueryOptions: structs.QueryOptions{
+					Region:    srv.Region(),
+					Namespace: ns1,
+					AuthToken: token,
+				},
+			},
+			expectVolIDs: []string{vol1.ID, vol2.ID},
+		},
+		{
+			name: "wildcard namespace",
+			req: &structs.HostVolumeListRequest{
+				QueryOptions: structs.QueryOptions{
+					Region:    srv.Region(),
+					Namespace: structs.AllNamespacesSentinel,
+					AuthToken: token,
+				},
+			},
+			expectVolIDs: []string{vol1.ID, vol2.ID, vol3.ID, vol4.ID},
+		},
+		{
+			name: "query by prefix",
+			req: &structs.HostVolumeListRequest{
+				QueryOptions: structs.QueryOptions{
+					Region:    srv.Region(),
+					Namespace: ns1,
+					AuthToken: token,
+					Prefix:    "foobar",
+				},
+			},
+			expectVolIDs: []string{vol1.ID},
+		},
+		{
+			name: "query by node",
+			req: &structs.HostVolumeListRequest{
+				NodeID: nodes[1].ID,
+				QueryOptions: structs.QueryOptions{
+					Region:    srv.Region(),
+					Namespace: structs.AllNamespacesSentinel,
+					AuthToken: token,
+				},
+			},
+			expectVolIDs: []string{vol2.ID, vol4.ID},
+		},
+		{
+			name: "query by node pool",
+			req: &structs.HostVolumeListRequest{
+				NodePool: "prod",
+				QueryOptions: structs.QueryOptions{
+					Region:    srv.Region(),
+					Namespace: structs.AllNamespacesSentinel,
+					AuthToken: token,
+				},
+			},
+			expectVolIDs: []string{vol3.ID},
+		},
+		{
+			name: "query by incompatible node ID and pool",
+			req: &structs.HostVolumeListRequest{
+				NodeID:   nodes[1].ID,
+				NodePool: "prod",
+				QueryOptions: structs.QueryOptions{
+					Region:    srv.Region(),
+					Namespace: structs.AllNamespacesSentinel,
+					AuthToken: token,
+				},
+			},
+			expectVolIDs: []string{},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			var resp structs.HostVolumeListResponse
+			err := msgpackrpc.CallWithCodec(codec, "HostVolume.List", tc.req, &resp)
+			must.NoError(t, err)
+			gotIDs := []string{}
+			for _, vol := range resp.Volumes {
+				gotIDs = append(gotIDs, vol.ID)
+			}
+			must.SliceContainsAll(t, tc.expectVolIDs, gotIDs,
+				must.Sprintf("got: %v", gotIDs))
+		})
+	}
+
+	t.Run("blocking query unblocks", func(t *testing.T) {
+
+		getReq := &structs.HostVolumeGetRequest{
+			ID: vol1.ID,
```

Comment:
I wasn't sure why this extra Get, but a little ways down I see it's to get the `Index` eh? could save a bit of cognitive load here

```suggestion
		// Get response will include the volume's Index to block on
		getReq := &structs.HostVolumeGetRequest{
			ID: vol1.ID,
```